File: src/main/java/com/everchanging/hierarchy/converter/TopEmployeeToEntityListConverter.java
package com.everchanging.hierarchy.converter;

import com.everchanging.hierarchy.entity.EmployeeEntity;
import com.everchanging.hierarchy.model.Employee;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.convert.converter.Converter;
import org.springframework.stereotype.Component;

import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

@Slf4j
@Component
public class TopEmployeeToEntityListConverter implements Converter<Employee, List<EmployeeEntity>> {
    @Override
    public List<EmployeeEntity> convert(@NotNull Employee topLevelEmployee) {
        log.debug("Convert from top topLevelEmployee {} to list of entities", topLevelEmployee);
        if (topLevelEmployee.getSupervisor() != null) {
            log.error("Received a non-top topLevelEmployee to convert into hierarchy map. {}", topLevelEmployee);
            throw new IllegalStateException();
        }

        List<EmployeeEntity> entities = new ArrayList<>();
        EmployeeEntity employeeEntity = new EmployeeEntity(topLevelEmployee.getName(), null);
        List<EmployeeEntity> subordinateEntities = convertSubordinatesToEntities(topLevelEmployee.getSubordinates(), employeeEntity);

        // Add top-level first then subordinates
        entities.add(employeeEntity);
        entities.addAll(subordinateEntities);

        log.debug("Returning entities {}", entities);
        return entities;
    }

    private List<EmployeeEntity> convertSubordinatesToEntities(Set<Employee> subordinates, EmployeeEntity supervisorEntity) {
        List<EmployeeEntity> entities = new ArrayList<>();
        for (Employee subordinate : subordinates) {
            EmployeeEntity subordinateEntity = new EmployeeEntity(subordinate.getName(), supervisorEntity);
            List<EmployeeEntity> subordinateEntities = convertSubordinatesToEntities(subordinate.getSubordinates(), subordinateEntity);

            // Add top-level first then subordinates
            entities.add(subordinateEntity);
            entities.addAll(subordinateEntities);
        }
        return entities;
    }
}

File: src/main/kotlin/com/everchanging/hierarchy/converter/TopEmployeeToEntityListConverter.kt
package com.everchanging.hierarchy.converter

import com.everchanging.hierarchy.entity.EmployeeEntity
import com.everchanging.hierarchy.model.Employee
import org.slf4j.LoggerFactory
import org.springframework.core.convert.converter.Converter
import org.springframework.stereotype.Component

@Component
class TopEmployeeToEntityListConverter : Converter<Employee, List<EmployeeEntity>> {
    companion object {
        private val log = LoggerFactory.getLogger(TopEmployeeToEntityListConverter::class.java)
    }

    override fun convert(topLevelEmployee: Employee): List<EmployeeEntity> {
        log.debug("Convert from top topLevelEmployee $topLevelEmployee to list of entities")
        if (topLevelEmployee.supervisor != null) {
            log.error("Received a non-top topLevelEmployee to convert into hierarchy map. $topLevelEmployee")
            throw IllegalStateException()
        }

        val entities = arrayListOf<EmployeeEntity>()
        val employeeEntity = EmployeeEntity(topLevelEmployee.name)
        val subordinateEntities = convertSubordinatesToEntities(topLevelEmployee.subordinates, employeeEntity)

        // Add top-level first then subordinates
        entities.add(employeeEntity)
        entities.addAll(subordinateEntities)

        log.debug("Returning entities $entities")
        return entities
    }

    private fun convertSubordinatesToEntities(subordinates: Set<Employee>, supervisorEntity: EmployeeEntity): List<EmployeeEntity> {
        val entities = arrayListOf<EmployeeEntity>()

        for (subordinate in subordinates) {
            val subordinateEntity = EmployeeEntity(subordinate.name, supervisorEntity)
            val subordinateEntities = convertSubordinatesToEntities(subordinate.subordinates, subordinateEntity)

            // Add top-level first then subordinates
            entities.add(subordinateEntity)
            entities.addAll(subordinateEntities)
        }
        return entities
    }
}